qudipy/control/voltage_pulses.py

```diff
@@ -0,0 +1,570 @@
+        # pulse duration
+        T = 2 * theta_rad / (omega * np.abs(delta_g_int))
+        
+    elif axis in ['X', 'Y']:
+        # determine the phase values during the pulse first
+        phis = np.full(num_val, 0.)
+        if axis.upper() == 'Y':
+            phis = np.full(num_val, 90)
+        if theta < 0:
+            phis = phis + 180
+            
+        rot_pulse.add_control_variable("phi", phis)
+        # Making the ESR pulse of the same shape as the g-factor pulse
+        # shifted by delta_g_0
+        # finding the coefficient first
+        
+        A = 1 / np.sqrt( (2 * np.pi / theta_rad) ** 2 - 1)
+        
+        # values of ESR field
+        omega_capital = A * omega * np.abs(delta_g_pulse - delta_g_0) / 2
+        b_rf = m_e / e * omega_capital 
+        
+        rot_pulse.add_control_variable("B_rf", b_rf)
+        
+        # unlike the case of ROTZ, the voltage here changes only on the 
+        # idling qubits so that they return to their initial state
+        # after the pulse
+        for i in idle_qubits:
+            rot_pulse.ctrl_pulses['delta_g_{ind}'.format(ind = i)] = delta_g_pulse
+            rot_pulse.ctrl_pulses['V_{ind}'.format(ind = i)] = v_pulse
+            
+        # pulse duration
+        alpha = theta_rad / (2 * np.pi)
+        T = 4 * np.pi / (omega * np.abs(delta_g_int)) *  np.sqrt(1 - alpha ** 2)
+    
+    else:
+        # *** ARBITRARY ROTATION ***
+        #create variables for Vector Components to speed up run-time
+        n_x = rotation_axis[0]
+        n_y = rotation_axis[1]
+        n_z = rotation_axis[2]
+        
+        # make sure that the rotation_axis is a Bloch Vector, if not, make it unitary
+        norm = np.sqrt(n_x**2 + n_y**2 + n_z**2)
+        if norm != 1:
+            n_x = n_x/norm
+            n_y = n_y/norm
+            n_z = n_z/norm
+        
+        def find_shape(active = True):
+            '''
+            @author: Madi
+            Calculates the shape function given that the qubit is
+            resonant (active) or non-resonant (inactive) for T = 1
+
+            Parameters
+            ----------------
+            active: boolean
+                is True when the qubit is active, False when the qubit is inactive
+
+            Returns
+            ---------------
+            List 
+                contains all of the values of the shape
+            '''
+            delta_g_diff = np.array(delta_g_pulse - delta_g_0)
+            if active:
+                param = (2 * n_z * theta_rad)/omega
+                return delta_g_diff / param
+            
+            else:
+                param = (2/omega) * np.sqrt(4 * np.pi**2 - theta_rad**2 * 
+                                                            (n_x**2 + n_y**2))
+                return delta_g_diff / param
+    
+        # Find Shape
+        if n_z == 1:
+            S = find_shape(active = True) #get shape
+        else:
+            S = find_shape(active = False) #get shape
```

Comment:
Can be replaced with a single line: 
```suggestion
            S = find_shape(active = (n_z == 1)) #get shape
```